src/server/gc_worker/compaction_filter.rs

```diff
@@ -154,133 +204,269 @@ struct WriteCompactionFilter {
+            let cf = db.cf_handle(CF_WRITE).unwrap();
+            let iter = DBIterator::new_cf(db.clone(), cf, ReadOptions::new());
+            filter.write_iter = Some(RocksEngineIterator::from_raw(iter));
+        }
+        filter
+    }
+
+    fn do_filter(
+        &mut self,
+        _start_level: usize,
+        key: &[u8],
+        sequence: u64,
+        value: &[u8],
+        value_type: CompactionFilterValueType,
+    ) -> Result<CompactionFilterDecision, String> {
+        self.meet_a_key_in_filter();
+
+        let (key_prefix, commit_ts) = split_ts(key)?;
+        if commit_ts > self.safe_point || value_type != CompactionFilterValueType::Value {
+            return Ok(CompactionFilterDecision::Keep);
+        }
+
+        self.versions += 1;
+        if self.key_prefix != key_prefix {
+            self.handle_delete_mark()?;
+            self.switch_key_metrics();
+            self.key_prefix.clear();
+            self.key_prefix.extend_from_slice(key_prefix);
+            self.remove_older = false;
+        }
+
+        let mut filtered = self.remove_older;
+        let write = parse_write(value)?;
+        if !self.remove_older {
+            match write.write_type {
+                WriteType::Rollback | WriteType::Lock => {
+                    self.mvcc_rollback_and_locks += 1;
+                    filtered = true;
+                }
+                WriteType::Put => self.remove_older = true,
+                WriteType::Delete => {
+                    self.remove_older = true;
+                    if self.is_bottommost_level {
+                        filtered = true;
+                        self.deleting_filtered = Some((key.to_vec(), Vec::new()));
+                    }
+                }
+            }
+        } else if let Some(deleting_filtered) = self.deleting_filtered.as_mut() {
+            deleting_filtered.1.push((commit_ts, sequence));
+        }
+
+        if !filtered {
+            return Ok(CompactionFilterDecision::Keep);
+        }
+        self.filtered += 1;
+        self.handle_filtered_write(write)?;
+        let decision = if self.remove_older {
+            debug_assert!(commit_ts > 0);
+            let prefix = Key::from_encoded_slice(key_prefix);
+            let skip_until = prefix.append_ts((commit_ts - 1).into()).into_encoded();
+            CompactionFilterDecision::RemoveAndSkipUntil(skip_until)
+        } else {
+            CompactionFilterDecision::Remove
+        };
+        Ok(decision)
+    }
+
+    fn meet_a_key_in_filter(&mut self) {
+        if self.is_bottommost_level && self.deleting_filtered.is_none() {
+            // The compaction hasn't met the next MVCC delete mark.
+            self.near_seek_distance += 1;
         }
     }
 
     // Do gc before a delete mark.
-    fn handle_delete_mark(&mut self, mark: &[u8]) {
-        let mut valid = self.write_iter_seek_to(mark);
+    fn handle_delete_mark(&mut self) -> Result<(), String> {
```

Comment:
add comment to explain why we need to use `write_iter` to iterate through all versions of the key before the deletion marker - because stale versions can appear in levels above the bottom level.